bench/irmin-pack/tree.ml

```diff
@@ -0,0 +1,398 @@
+    Store.Commit.of_hash repo prev_commit >>= function
+    | None -> Lwt.fail_with "commit not found"
+    | Some commit ->
+        let tree = Store.Commit.tree commit in
+        f tree >>= fun tree ->
+        Store.Commit.v repo ~info:(info ()) ~parents:[ prev_commit ] tree
+
+  let add_commits repo ncommits f () =
+    init_commit repo >>= fun c ->
+    let rec aux c i =
+      if i >= ncommits then Lwt.return c
+      else
+        checkout_and_commit repo (Store.Commit.hash c) f >>= fun c' ->
+        aux c' (i + 1)
+    in
+    aux c 0 >|= ignore
+
+  let run mode config =
+    let conf = Irmin_pack.config ~readonly:false ~fresh:true config.root in
+    Store.Repo.v conf >>= fun repo ->
+    (match mode with
+    | `Large ->
+        Trees.add_large_trees config.width config.nlarge_trees
+        |> add_commits repo config.ncommits
+    | `Chains ->
+        Trees.add_chain_trees config.depth config.nchain_trees
+        |> add_commits repo config.ncommits
+    | `Tezos_log -> Trees_log.add_commits repo
+    | `All -> assert false)
+    |> Benchmark.run config
+    >>= fun result ->
+    Store.Repo.close repo >|= fun () -> result
+end
+
+module Bench_inodes_32 = Bench_suite (Conf)
+
+module Bench_inodes_2 = Bench_suite (struct
+  let entries = 2
+  let stable_hash = 5
+end)
+
+let init root =
+  FSHelper.rm_dir root;
+  reset_stats ()
+
+let pp_config mode fmt config =
+  match mode with
+  | `Tezos_log ->
+      let stats = Bench_inodes_32.Trees_log.get_stats () in
+      Format.fprintf fmt
+        "Tezos_log mode: %d commits (%d adds; %d removes; %d finds; %d mems; \
+         %d mem_tree; %d copies)"
+        config.ncommits_log stats.adds stats.removes stats.finds stats.mems
+        stats.mem_tree stats.copies
+  | `Chains ->
+      Format.fprintf fmt
+        "Chain trees mode: %d commits, each consisting of %d chains of depth %d"
+        config.ncommits config.nchain_trees config.depth
+  | `Large ->
+      Format.fprintf fmt
+        "Large trees mode: %d commits, each consisting of %d large trees of %d \
+         entries"
+        config.ncommits config.nlarge_trees config.width
+  | `All -> assert false
+
+let main ncommits ncommits_log operations_file mode depth width nchain_trees
+    nlarge_trees =
+  Printexc.record_backtrace true;
+  let config =
+    {
+      ncommits;
+      ncommits_log;
+      operations_file;
+      root = "test-bench";
+      mode;
+      depth;
+      width;
+      nchain_trees;
+      nlarge_trees;
+    }
+  in
+  let run_benchmarks () =
+    (match config.mode with
+    | `All -> [ `Tezos_log; `Chains; `Large ]
+    | (`Chains | `Large | `Tezos_log) as mode -> [ mode ])
+    |> Lwt_list.map_s (fun mode ->
+           init config.root;
+           if mode = `Tezos_log then (
+             Parse_log.prepare_log config.operations_file;
+             let n =
+               Parse_log.populate_array "tezos_log.json" config.ncommits_log
+             in
+             config.ncommits_log <- n;
```

Comment:
We could pass a new config in via `{ config with ncommits_log = n }` and avoid mutating state that could potentially be read by other benchmarks.